core/src/main/java/org/geysermc/geyser/session/PendingMicrosoftAuthentication.java

```diff
@@ -0,0 +1,136 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Pending Microsoft authentication task cache.
+ * It permits user to exit the server while they authorize Geyser to access their Microsoft account.
+ */
+public class PendingMicrosoftAuthentication {
+    private final LoadingCache<String, AuthenticationTask> authentications;
+
+    public PendingMicrosoftAuthentication(int timeoutSeconds) {
+        this.authentications = CacheBuilder.newBuilder()
+                .expireAfterWrite(Duration.ofSeconds(timeoutSeconds))
+                .build(new CacheLoader<>() {
+                    @Override
+                    public AuthenticationTask load(@NonNull String userKey) {
+                        return new AuthenticationTask(userKey, timeoutSeconds * 1000L);
+                    }
+                });
+    }
+
+    @SneakyThrows(ExecutionException.class)
+    public AuthenticationTask getOrCreateTask(@NonNull String userKey) {
+        return authentications.get(userKey);
+    }
+
+    @FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+    public static class AuthenticationTask {
+        MsaAuthenticationService msaAuthenticationService = new MsaAuthenticationService(GeyserImpl.OAUTH_CLIENT_ID);
+        String userKey;
+        long timeoutMs;
+
+        @Getter
+        CompletableFuture<MsaAuthenticationService.MsCodeResponse> code;
+        @Getter
+        CompletableFuture<MsaAuthenticationService> authentication;
+
+        private AuthenticationTask(String userKey, long timeoutMs) {
+            this.userKey = userKey;
+            this.timeoutMs = timeoutMs;
+
+            // Request the code
+            this.code = CompletableFuture.supplyAsync(this::tryGetCode);
+
+            // Once the code is received, continuously try to request the access token, profile, etc
+            this.authentication = code.thenApply((code) -> tryLoginContinuously());
+        }
+
+        private MsaAuthenticationService.MsCodeResponse tryGetCode() throws CompletionException {
+            try {
+                return msaAuthenticationService.getAuthCode();
+            } catch (RequestException e) {
+                throw new CompletionException(e);
+            }
+        }
+
+        private MsaAuthenticationService tryLoginContinuously() throws CompletionException {
+            try {
+                long startTime = System.currentTimeMillis();
+                while (true) {
+                    try {
+                        msaAuthenticationService.login();
+                    } catch (AuthPendingException e) {
+                        long deltaTime = System.currentTimeMillis() - startTime;
+                        if (deltaTime > timeoutMs) {
+                            throw new TaskTimeoutException();
+                        }
+                        //noinspection BusyWait
+                        Thread.sleep(1000);
```

Comment:
How well does the common fork join pool (which I believe CompletableFuture defaults to running under) handle thread sleeping like this? If it holds up a thread and prevents it from being re-used I don't think this is good.

Reply:
You're right. Next check should be scheduled.